cpp/src/parquet/encoding_test.cc

```diff
@@ -867,158 +902,71 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
-}
+template <typename Type>
+class TestByteStreamSplitEncoding : public TestEncodingBase<Type> {
+ public:
+  typedef typename Type::c_type T;
+  static constexpr int TYPE = Type::type_num;
 
-// Check that the encoder can handle input with one element.
-TEST(ByteStreamSplitEncodeDecode, EncodeOneLenInput) {
-  const float value = 1.0f;
-  TestByteStreamSplitRoundTrip<FloatType>(&value, 1);
-}
+  virtual void CheckRoundtrip() {
+    auto encoder =
+        MakeTypedEncoder<Type>(Encoding::BYTE_STREAM_SPLIT, false, descr_.get());
+    auto decoder = MakeTypedDecoder<Type>(Encoding::BYTE_STREAM_SPLIT, descr_.get());
+    encoder->Put(draws_, num_values_);
+    encode_buffer_ = encoder->FlushValues();
 
-// Check that the decoder can handle empty input.
-TEST(ByteStreamSplitEncodeDecode, DecodeZeroLenInput) {
-  std::unique_ptr<TypedDecoder<FloatType>> decoder =
-      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
-  decoder->SetData(0, NULL, 0);
-  ASSERT_EQ(0U, decoder->Decode(NULL, 0));
-}
+    {
+      decoder->SetData(num_values_, encode_buffer_->data(),
+                       static_cast<int>(encode_buffer_->size()));
+      int values_decoded = decoder->Decode(decode_buf_, num_values_);
+      ASSERT_EQ(num_values_, values_decoded);
+      ASSERT_NO_FATAL_FAILURE(VerifyResults<T>(decode_buf_, draws_, num_values_));
+    }
 
-TEST(ByteStreamSplitEncodeDecode, DecodeOneLenInput) {
-  const uint8_t data[] = {0x47U, 0x24U, 0xa7U, 0x44U};
-  TestByteStreamSplitDecodePath<FloatType>(
-      data, 4, reinterpret_cast<const float*>(&data[0]), 1, false);
-}
+    {
+      // Try again but with a small step.
+      decoder->SetData(num_values_, encode_buffer_->data(),
+                       static_cast<int>(encode_buffer_->size()));
+      int step = 131;
+      int remaining = num_values_;
+      for (int i = 0; i < num_values_; i += step) {
+        int num_decoded = decoder->Decode(decode_buf_, step);
+        if (remaining >= step) {
+          ASSERT_EQ(step, num_decoded);
+        } else {
+          ASSERT_EQ(num_decoded, remaining);
+        }
+        ASSERT_NO_FATAL_FAILURE(VerifyResults<T>(decode_buf_, &draws_[i], num_decoded));
+        remaining -= num_decoded;
+      }
+    }
 
-// Check that requesting to decode more elements than is available in the storage
-// of the decoder works correctly.
-TEST(ByteStreamSplitEncodeDecode, DecodeLargerPortion) {
-  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U, 0x11U, 0x22U, 0x33U, 0x44U,
-                          0xAAU, 0xBBU, 0xCCU, 0xDDU, 0x55U, 0x66U, 0x77U, 0x88U};
-  const uint64_t expected_output[2] = {0x7755CCAA331137DEULL, 0x8866DDBB442213C0ULL};
-  TestByteStreamSplitDecodePath<DoubleType>(
-      data, sizeof(data), reinterpret_cast<const double*>(&expected_output[0]), 2, true);
-}
+    {
+      std::vector<uint8_t> valid_bits(arrow::BitUtil::BytesForBits(num_values_) + 1, 255);
+      ASSERT_NO_THROW(encoder->PutSpaced(draws_, num_values_, valid_bits.data(), 0));
```

Comment:
Also, the `PutSpaced` test would be more interesting with a non-trivial validity bitmap (i.e. not all 1's).

Reply:
Changed so that every fifth element is only enabled for adding.